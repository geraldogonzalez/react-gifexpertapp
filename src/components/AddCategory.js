import React, { useState } from 'react';
import PropTypes from 'prop-types';

const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');

    const inputHandleChange = e => {
        setInputValue(e.target.value);
    }
    
    const handleSubmit = e => {
        e.preventDefault();
        
        if(inputValue.trim().length > 0){
            setCategories(cats => [ inputValue, ...cats ]);
            setInputValue('');
        }

    }

    return (
        <form onSubmit={ handleSubmit } >

            <p>{inputValue}</p>
            <input 
                type="text"
                value={ inputValue }
                onChange={ inputHandleChange }
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}

export default AddCategory;