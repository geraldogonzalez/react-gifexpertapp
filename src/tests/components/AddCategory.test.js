import React from 'react';
import '@testing-library/jest-dom';

import { shallow } from 'enzyme';
import AddCategory from '../../components/AddCategory';

describe('Pruebas sobre el componente <AddCategory />', () => {
   
    const setCategories = jest.fn();
    const setInputValue = jest.fn()

    let wrapper = shallow( <AddCategory setCategories={setCategories} /> );

    beforeEach( () => {

        jest.clearAllMocks();

        wrapper = shallow( <AddCategory setCategories={setCategories} /> );

    });

    test('Debe recibir una función como props', () => {

        expect( wrapper ).toMatchSnapshot();

    })

    test('Debe cambiar la caja de texto', () => {
        
        const input = wrapper.find('input');
        const value = 'Hola mundo';
        
        input.simulate('change', { target: { value } } ) 

        expect( wrapper.find('p').text().trim() ).toBe( value );

    })

    test('NO debe postear la información con submit', () => {
        
        const form = wrapper.find('form');

        form.simulate('submit', { preventDefault(){} });

        expect( setCategories ).not.toHaveBeenCalled();

    })

    test('Debe llamar el setCategories y limpiar la caja de texto', () => {
        
        // 1. Simular el inputChange. 
        const input = wrapper.find('input');
        
        input.simulate('change', { target: { value: 'Hola' } } )

        // 2. Simular el submit.
        
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        // 3. setCategories se debe de haber llamado.

        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1);
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );
        
        // 4. El valor del input debe ser igual a un string vacío ('')
        expect( input.prop('value') ).toBe('');

    })
    
    

});
