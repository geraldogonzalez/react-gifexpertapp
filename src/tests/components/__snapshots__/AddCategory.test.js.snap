// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre el componente <AddCategory /> Debe recibir una funci√≥n como props 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
